File: 47. Permutations II.py
# https://leetcode.com/problems/permutations-ii/

# Given a collection of numbers, nums, that might contain duplicates, return
# all possible unique permutations in any order.

###############################################################################

# all unique combos -> sort, dfs -> add child that is not used
# is_used = [] for recording
# avoid duplicate: i) not used AND ii) first child OR (!= prev) OR (== prev, but prev is used)

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        if not nums or len(nums) == 0: return []
        
        nums.sort()
        self.ans = []
        self.holder = []
        self.used = set()
        self.nums = nums
        self.dfs()
        return self.ans
    
    def dfs(self):
        if len(self.holder) == len(self.nums):
            self.ans.append(self.holder.copy())
            return
        
        # generate child
        for i in range(len(self.nums)):
            # i) not used
            if i not in self.used:
                # ii) first child OR (!= prev) OR (== prev, but prev is used)
                if i == 0 or self.nums[i] != self.nums[i-1] or i - 1 in self.used:
                    self.holder.append(self.nums[i])
                    self.used.add(i)
                    
                    self.dfs()
                    
                    self.holder.pop()
                    self.used.remove(i)

File: 47. Permutations II.cpp
// https://leetcode.com/problems/permutations-ii/

// Given a collection of numbers, nums, that might contain duplicates, return all
// possible unique permutations in any order.

////////////////////////////////////////////////////////////////////////////////

// dfs, sort first
// avoid duplicate: first child, or != prev, or == prev but prev is used
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        used = vector<bool>(nums.size(), false);
        dfs(nums);
        return ans;
    }
private:
    vector<bool> used;
    vector<int> holder;
    vector<vector<int>> ans;
    void dfs(const vector<int>& nums) {
        if (holder.size() == nums.size()) {
            ans.push_back(holder); return;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (used[i]) continue;
            // first child, or != prev, or == prev but prev is used
            if (i == 0 || nums[i] != nums[i-1] || used[i-1]) {
                holder.push_back(nums[i]);
                used[i] = true;
                dfs(nums);
                holder.pop_back();
                used[i] = false;
            }
        }
        return;
    }
};
